git/odb/pack/index.go

```diff
@@ -28,8 +28,97 @@ func (i *Index) Count() int {
 	return int(i.fanout[255])
 }
 
+// Entry returns an entry containing the offset of a given SHA1 "name".
+//
+// Entry operates in O(log(n))-time in the worst case, where "n" is the number
+// of objects that begin with the first byte of "name".
+//
+// If the entry cannot be found, (nil, nil) will be returned. If there was an
+// error searching for or parsing an entry, it will be returned as (nil, err).
+//
+// Otherwise, (entry, nil) will be returned.
+func (i *Index) Entry(name []byte) (*IndexEntry, error) {
+	var last *bounds
+	bounds := i.bounds(name)
+
+	for bounds.Left() < bounds.Right() {
+		if last.Equal(bounds) {
+			// If the bounds are unchanged, that means either that
+			// the object does not exist in the packfile, or the
+			// fanout table is corrupt.
+			//
+			// Either way, we won't be able to find the object.
+			// Return immediately to prevent infinite looping.
+			return nil, nil
+		}
+		last = bounds
+
+		// Find the midpoint between the upper and lower bounds.
+		mid := (bounds.Left() + bounds.Right()) / 2
```

Comment:
I don't know how Go handles integer overflow, but this is subject to overflow when the size of the bounds approaches the capacity of your left/right integer types. One overflow-safe way is `bounds.Left() + (bounds.Right() - bounds.Left()) / 2`. Since you're using 64-bit integers it seems pretty unlikely in practice, though, and you can probably ignore it.